Transform discard binding

JLHwung opened the pull request:
<!--
Before making a PR, please read our contributing guidelines
https://github.com/babel/babel/blob/main/CONTRIBUTING.md

Please note that the Babel Team requires two approvals before merging most PRs.

For issue references: Add a comma-separated list of a [closing word](https://help.github.com/articles/closing-issues-via-commit-messages/) followed by the ticket number fixed by the PR. (it should be underlined in the preview if done correctly)

If you are making a change that should have a docs update: submit another PR to https://github.com/babel/website
-->

| Q                        | A <!--(Can use an emoji 👍) -->
| ------------------------ | ---
| Fixed Issues?            | Transform discard binding
| Patch: Bug Fix?          |
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | https://github.com/babel/website/pull/3095
| Any Dependency Changes?  |
| License                  | MIT

<!-- Describe your changes below in as much detail as possible -->
This PR implements the [Discard Binding Proposal](https://tc39.es/proposal-discard-binding) transform.

- [x] restructure `index.js`
- [x] Expose `syntaxType` as a transform plugin option
- [x] Remove trailing void patterns in function declarations when the `ignoreFunctionLength` assumption is enabled
- [x] More tests
- [x] Fix object-rest-spread transform (https://github.com/babel/babel/pull/17281)
- [x] Remove `VoidPattern` from `LVal`

JLHwung commented on packages/babel-plugin-proposal-discard-binding/src/utils.ts:
```diff
@@ -0,0 +1,146 @@
+  return !iterateVoidPatternInVariableDeclaration(path).next().done;
+}
+
+export function splitNamedDeclarationAsVarAndExport(
```
This helper is also used in the `destructuring` and `object-rest-spread` transform. Do you think we should extract them into one `helper-destructuring-transform` package and reuse them in transforms? Or a mega `helper-create-destructuring-transform` package that lowers to different compilation targets?

nicolo-ribaudo replied:
Maybe let's start wither with a small helper package, or by exporting it from `destructuring` and then making this plugin depend on it.

nicolo-ribaudo reviewed (commented):
I left some comments, but it looks overall good 👍 Really nice to see how generators can make traversals look simpler.

JLHwung commented on packages/babel-plugin-proposal-destructuring-private/src/util.ts:
```diff
@@ -421,6 +421,13 @@ export function* transformPrivateKeyDestructuring(
             }
             // An object rest element must not contain a private key
             const property = properties[index] as t.ObjectProperty;
+            if (property.value.type === "VoidPattern") {
```
Yes, otherwise we will have to add a class visitor and traverse every pattern parents, which is a bit overkill, so here we just let the private destructuring handle void patterns in array pattern.

JLHwung commented on packages/babel-plugin-proposal-discard-binding/test/fixtures/integration-explicit-resource-management/basic/output.js:
```diff
@@ -0,0 +1,8 @@
+try {
+  var _usingCtx = babelHelpers.usingCtx();
+  const _ = _usingCtx.u(null);
```
Yes. The named evaluation transform will also benefit from the awareness of the void pattern: since in that case we should not inject `setFunctionName` call at all. I am considering between these three approaches:

1. let ERM import this plugin and handle the void pattern transform there
2. When this plugin handles `void` within `using`, transform it to Identifier, and inject a field indicating that it was a void pattern to the AST extra.
3. Inject the void pattern transform to PluginPass, and read/evaluate it from the ERM transform. Essentially this approach is adding an implicit dependency of void pattern for the ERM transform.

Feel free to share if you have another preferred approach.

nicolo-ribaudo replied:
I slightly prefer (2), adding maybe a `.extra.ignoreFunctionName` field to the identifier or something like that. (1) would also be fine. I think none of the option is particularly clean :sweat_smile:

JLHwung replied:
You are right. I am now settled in the 4th approach: parenthesize the rhs class if `using void` is seen and then transform `void` as usual. With a small fix in the generator, this approach should work with both the ERM syntax and the ERM transform plugins.

JLHwung replied:
The prefix will be injected only when the rhs is an anonymous class expression with a static dispose method. 

I assume the typical use case is to create a resource with instance dispose method and initialize it from the class. So the prefix will not kick in for most use cases.

liuxingbaoyu commented on .prettierignore:
```diff
@@ -8,6 +8,7 @@ packages/babel-standalone/examples/**
 packages/babel-parser/test/expressions/**
 scripts/integration-tests/fixtures/**
 packages/babel-parser/test/schema.json
+packages/babel-plugin-proposal-discard-binding/test/scripts/check-transform-scope-integrity.cjs
```
Is there any reason for this?

JLHwung commented:
I have pushed some new changes:

- Removed `VoidPattern` from `LVal`: `void = rhs` is never valid. Previously we add `VoidPattern` to `LVal` because `VariableDeclarator` requires `id: LVal`: This has been fixed in Babel 8. So in this PR we just broaden `id` to `LVal | VoidPattern`.
- The new typing changes reveal a bug in the destructuring-transform: Because `VoidPattern` is a `Pattern` that the destructuring-transform can not handle, it will render into an endless recursion. This is a rare edge case that requires enabling both ERM and DiscardBinding but apply the destructuring transform. Anyway we have fixed that as well.
- Typing updates per the typing changes

nicolo-ribaudo commented on packages/babel-plugin-proposal-discard-binding/src/utils.ts:
```diff
@@ -0,0 +1,165 @@
+
+// https://tc39.es/ecma262/#sec-isanonymousfunctiondefinition
+// We don't test anonymous function / arrow function because they must not be disposable
+export function isAnonymousFunctionDefinition(
```
I don't think we need this check. There is no good reason for anybody to write `using void = () => {}`, but if they do we might as well inject (`0,`) and avoid having two different code paths for classes and functions.

JLHwung replied:
Well, if they do they will get a runtime error for sure. Added the function / arrow function check.

nicolo-ribaudo reviewed (approved):
Minor comment, but overall looks good.
